File: java/src/main/java/com/jade/programmers/practice/level1/P_소수_찾기.java
package com.jade.programmers.practice.level1;


/**
 * 문제 주소 - https://programmers.co.kr/learn/courses/30/lessons/12921
 * 문제 설명
 * 1부터 입력받은 숫자 n 사이에 있는 소수의 개수를 반환하는 함수, solution을 만들어 보세요.
 * <p>
 * 소수는 1과 자기 자신으로만 나누어지는 수를 의미합니다.
 * (1은 소수가 아닙니다.)
 * <p>
 * 제한 조건
 * n은 2이상 1000000이하의 자연수입니다.
 */
public class P_소수_찾기 {
    public int solution(int n) {
        boolean[] notPrimeNumbers = new boolean[n + 1];
        notPrimeNumbers[0] = true;
        notPrimeNumbers[1] = true;

        for (int i = 2; i < n; i++) {
            if (notPrimeNumbers[i]) {
                continue;
            }

            for (int j = i + i; j <= n; j += i) {
                notPrimeNumbers[j] = true;
            }
        }
        int count = 0;
        for (boolean b : notPrimeNumbers) {
            if (!b) {
                count++;
            }
        }
        return count;
    }
}

File: python/programmers/level2/P_소수_찾기.py
# https://programmers.co.kr/learn/courses/30/lessons/42839

from itertools import permutations


def isPrimeNumber(n):
    if n < 2:
        return False
    for i in range(2, n):
        if n % i == 0:
            return False
    return True


def solution(numbers):
    answer = 0
    l = [str(i) for i in numbers]
    s = {0}
    for i in range(len(numbers) + 1):
        p = list(map(''.join, permutations(l, i)))
        if p == ['']:
            continue
        s.update(list(map(int, p)))
    for i in s:
        if isPrimeNumber(int(i)):
            print
            answer += 1

    return answer

if __name__ == '__main__':
    print(solution("17"))
    print(solution("011"))
